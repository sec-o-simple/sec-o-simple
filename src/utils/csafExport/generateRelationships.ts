import {
  CSAFRelationship,
  TRelationship,
} from '@/routes/products/types/tRelationship'
import { PidGenerator } from './pidGenerator'
import { uid } from 'uid'

export default function generateRelationships(
  relationships: TRelationship[],
  pidGenerator: PidGenerator,
): CSAFRelationship[] {
  const csafRelationships = [] as CSAFRelationship[]
  const ridGenerator = new PidGenerator()
  ridGenerator.prefix = 'CSAFRID'

  relationships.forEach((relationship) => {
    relationship.product1VersionIds.forEach((sourceVersion) => {
      relationship.product2VersionIds.forEach((targetVersion) => {
        csafRelationships.push({
          category: relationship.category,
          product_reference: pidGenerator.getPid(sourceVersion),
          relates_to_product_reference: pidGenerator.getPid(targetVersion),
          full_product_name: {
            name: relationship.name,
            product_id: ridGenerator.getPid(uid()),
          },
        })
      })
    })
  })

  return csafRelationships
}
