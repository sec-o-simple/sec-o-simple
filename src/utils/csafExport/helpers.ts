export function getFilename(trackingId: string): string {
  return trackingId.toLowerCase().replaceAll(/[^+\-a-z0-9]+/g, '_')
}
