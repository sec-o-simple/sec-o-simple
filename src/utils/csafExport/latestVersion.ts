import { TRevisionHistoryEntry } from "@/routes/document-information/types/tRevisionHistoryEntry";
import { compareVersions } from "../version";

/**
 * Compares two Date objects and returns the difference in milliseconds.
 * Used for sorting dates in chronological order.
 * 
 * @param a - The first date to compare
 * @param z - The second date to compare
 * @returns A negative number if a is before z, positive if a is after z, or 0 if they are equal
 */
function compareZonedDateTimes(a: Date, z: Date): number {
    return a.getTime() - z.getTime()
}

/**
 * Retrieves the version number of the most recent revision from a revision history array.
 * 
 * The function sorts the revision history by date (most recent first) and then by version number
 * as a secondary sort criteria. It returns the version number of the first entry after sorting.
 * 
 * @param history - Array of revision history entries containing date and version information.
 *                  Each entry must have a `date` property (string) and a `number` property (string).
 * @returns The version number of the latest revision as a string
 * 
 * @throws {Error} Throws if the history array is empty.
 * 
 * @example
 * ```typescript
 * const history = [
 *   { date: '2024-01-01', number: '1.0.0' },
 *   { date: '2024-01-15', number: '1.1.0' },
 *   { date: '2024-01-10', number: '1.0.1' }
 * ];
 * const latest = retrieveLatestVersion(history); // Returns '1.1.0'
 * ```
 * 
 * @example
 * ```typescript
 * // When dates are the same, version comparison is used as secondary sort
 * const sameDate = [
 *   { date: '2024-01-01', number: '1.0.0' },
 *   { date: '2024-01-01', number: '1.1.0' }
 * ];
 * const latest = retrieveLatestVersion(sameDate); // Returns '1.1.0'
 * ```
 */
export function retrieveLatestVersion(history: TRevisionHistoryEntry[]): string {
    if (history.length === 0) {
        throw new Error("Revision history is empty, cannot retrieve latest version.");
    }

    const sorted = history.slice()
        .sort(
            (a, z) =>
                compareZonedDateTimes(
                    new Date(z.date),
                    new Date(a.date),
                ) || compareVersions(a.number, z.number)
        )[0].number

    return sorted
}