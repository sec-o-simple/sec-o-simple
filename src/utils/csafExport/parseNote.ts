import { TNote, TNoteCategory } from '@/routes/shared/NotesList'

export type TParsedNote = {
  category: TNoteCategory
  text: string
  title: string
}

export function parseNote(note: TNote): TParsedNote {
  return {
    category: note.category,
    text: note.content,
    title: note.title,
  }
}
