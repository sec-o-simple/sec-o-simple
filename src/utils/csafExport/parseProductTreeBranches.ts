import { getFamilyChain } from '@/routes/products/ProductFamily'
import {
  TProductFamily,
  TProductIdentificationHelper,
  TProductTreeBranch,
  TProductTreeBranchCategory,
} from '@/routes/products/types/tProductTreeBranch'

export type TParsedProductTreeBranch = {
  category: TProductTreeBranchCategory
  name: string
  product?: {
    name: string
    product_id: string
    product_identification_helper?: TProductIdentificationHelper
  }
  branches?: TParsedProductTreeBranch[]
}

export function parseProductTreeBranches(
  branches: TProductTreeBranch[],
  families: TProductFamily[],
  getFullProductName: (id: string) => string,
): TParsedProductTreeBranch[] {
  // Helper function to find a family by ID
  const findFamilyById = (familyId: string): TProductFamily | null => {
    return families.find((family) => family.id === familyId) || null
  }

  // Helper function to build family tree branches from a family chain
  const buildFamilyBranches = (
    familyChain: TProductFamily[],
    productBranch: TParsedProductTreeBranch,
  ): TParsedProductTreeBranch => {
    // Start from the deepest family and work up
    let result = productBranch

    for (let i = familyChain.length - 1; i >= 0; i--) {
      const family = familyChain[i]
      result = {
        category: 'product_family',
        name: family.name,
        branches: [result],
      }
    }

    return result
  }

  return branches.map((branch) => {
    const pbObj: TParsedProductTreeBranch = {
      category: branch.category,
      name: branch.name,
    }

    // Handle sub-branches recursively if they exist
    if (branch.subBranches.length > 0) {
      pbObj['branches'] = parseProductTreeBranches(
        branch.subBranches,
        families,
        getFullProductName,
      )
    }

    // Only add product property for product_version branches (leaf nodes in the hierarchy)
    // The hierarchy is: vendor → product-family (optional, can be nested) → product → version
    if (branch.category === 'product_version') {
      pbObj['product'] = {
        name: branch.productName ?? getFullProductName(branch.id),
        product_id: branch.id,
        product_identification_helper: branch.identificationHelper,
      }
    }

    // If this is a product_name branch with a familyId, we need to integrate the family hierarchy
    if (branch.category === 'product_name' && branch.familyId) {
      const family = findFamilyById(branch.familyId)

      if (family) {
        const familyChain = getFamilyChain(family)

        if (familyChain.length > 0) {
          // Build the family tree structure with the product at the end
          const productBranch: TParsedProductTreeBranch = {
            category: branch.category,
            name: branch.name,
            branches: pbObj.branches, // Include product versions
          }

          // Wrap the product in the family hierarchy
          return buildFamilyBranches(familyChain, productBranch)
        }
      }
    }

    return pbObj
  })
}
