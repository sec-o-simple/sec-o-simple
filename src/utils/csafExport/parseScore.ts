import { TVulnerabilityScore } from '@/routes/vulnerabilities/types/tVulnerabilityScore'
import { calculateBaseScore, calculateQualScore } from 'cvss4'
import { PidGenerator } from './pidGenerator'

export default function parseScore(
  score: TVulnerabilityScore,
  pidGenerator: PidGenerator,
) {
  let baseScore = 0
  let baseSeverity = ''

  try {
    baseScore = calculateBaseScore(score.vectorString)
    baseSeverity = calculateQualScore(baseScore).toUpperCase()
  } catch {
    // If the score is invalid, we leave baseScore and baseSeverity as defaults
    // as there will be errors already in the vectorString
  }

  const key = score.cvssVersion === '3.1' ? 'cvss_v3' : 'cvss_v4'

  return {
    [key]: {
      version: score.cvssVersion,
      vectorString: score.vectorString,
      baseScore,
      baseSeverity,
    },
    products: score.productIds.map((id) => pidGenerator.getPid(id)),
  }
}
