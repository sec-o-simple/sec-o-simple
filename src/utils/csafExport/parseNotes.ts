import { TDocumentInformation } from '@/routes/document-information/types/tDocumentInformation'
import { TProductTreeBranch } from '@/routes/products/types/tProductTreeBranch'
import { TDocumentStore } from '../useDocumentStore'
import { parseNote, TParsedNote } from './parseNote'

function extractAllProducts(
  products: TProductTreeBranch[],
): TProductTreeBranch[] {
  return products?.flatMap((vendor) => {
    return vendor.subBranches?.flatMap((product) => product)
  })
}

export function parseNotes(documentStore: TDocumentStore): TParsedNote[] {
  const documentInformation: TDocumentInformation =
    documentStore.documentInformation
  const notes = documentInformation.notes.map(parseNote)

  const productNotes = extractAllProducts(Object.values(documentStore.products))
    ?.filter((p) => p.description.length)
    .map((product) => ({
      category: 'description',
      text: product.description || '',
      title:
        documentInformation.lang === 'de'
          ? `Produktbeschreibung f√ºr ${product.name}`
          : `Product description for ${product.name}`,
    }))

  return [...notes, ...productNotes] as TParsedNote[]
}
