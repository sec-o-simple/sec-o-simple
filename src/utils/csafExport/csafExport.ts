import { calculateBaseScore, calculateQualScore } from 'cvss4'
import { download } from '../download'
import useDocumentStore, { TDocumentStore } from '../useDocumentStore'
import generateRelationships from './generateRelationships'
import { getFilename } from './helpers'
import { retrieveLatestVersion } from './latestVersion'
import { parseNote } from './parseNote'
import { parseProductTreeBranches } from './parseProductTreeBranches'
import { PidGenerator } from './pidGenerator'

export type TCSAFDocument = ReturnType<typeof createCSAFDocument>

export function createCSAFDocument(documentStore: TDocumentStore) {
  const pidGenerator = new PidGenerator()
  const currentDate = new Date().toISOString()

  const csafDocument = {
    document: {
      category: 'csaf_security_advisory',
      csaf_version: '2.0',
      tracking: {
        generator: {
          date: currentDate,
          engine: {
            version: '0.0.1',
            name: 'Sec-O-Simple',
          },
        },
        current_release_date: currentDate,
        initial_release_date: currentDate,
        revision_history: documentStore.documentInformation.revisionHistory.map(
          (entry) => ({
            date: entry.date,
            number: entry.number,
            summary: entry.summary,
          }),
        ),
        status: documentStore.documentInformation.status,
        version: documentStore.documentInformation.revisionHistory.length
          ? retrieveLatestVersion(
              documentStore.documentInformation.revisionHistory,
            )
          : '1',
        id: documentStore.documentInformation.id,
      },
      lang: documentStore.documentInformation.language,
      title: documentStore.documentInformation.title,
      publisher: {
        category: documentStore.documentInformation.publisher.category,
        contact_details:
          documentStore.documentInformation.publisher.contactDetails,
        issuing_authority:
          documentStore.documentInformation.publisher.issuingAuthority ||
          undefined,
        name: documentStore.documentInformation.publisher.name,
        namespace: documentStore.documentInformation.publisher.namespace,
      },
      notes: documentStore.documentInformation.notes.map(parseNote),
      references: documentStore.documentInformation.references.map(
        (reference) => ({
          summary: reference.summary,
          url: reference.url,
          category: reference.category,
        }),
      ),
    },
    product_tree: {
      branches: parseProductTreeBranches(
        Object.values(documentStore.products),
        pidGenerator,
      ),
      relationships: generateRelationships(
        documentStore.relationships,
        pidGenerator,
      ),
    },
    vulnerabilities: Object.values(documentStore.vulnerabilities).map(
      (vulnerability) => ({
        cve: vulnerability.cve || undefined,
        title: vulnerability.title,
        cwe: vulnerability.cwe
          ? {
              id: vulnerability.cwe.id,
              name: vulnerability.cwe.name,
            }
          : undefined,
        notes: vulnerability.notes.map(parseNote),
        product_status: {
          known_affected: vulnerability.products.map((p) =>
            pidGenerator.getPid(p.firstAffectedVersionId),
          ),
          fixed: vulnerability.products.map((p) =>
            pidGenerator.getPid(p.firstFixedVersionId),
          ),
        },
        remediations: vulnerability.remediations.map((remediation) => ({
          category: remediation.category,
          date: remediation.date,
          details: remediation.details,
          url: remediation.url,
          product_ids: remediation.productIds.map((id) =>
            pidGenerator.getPid(id),
          ),
        })),
        scores: vulnerability.scores.map((score) => {
          let baseScore = 0
          let baseSeverity = ''

          try {
            baseScore = calculateBaseScore(score.vectorString)
            baseSeverity = calculateQualScore(baseScore).toUpperCase()
          } catch {
            // If the score is invalid, we leave baseScore and baseSeverity as defaults
            // as there will be errors already in the vectorString
          }

          return {
            ['cvss_v3']: {
              version: '3.1',
              vectorString: score.vectorString,
              baseScore,
              baseSeverity,
            },
            products: score.productIds.map((id) => pidGenerator.getPid(id)),
          }
        }),
      }),
    ),
  }

  return csafDocument
}

export function useCSAFExport() {
  const documentStore = useDocumentStore()

  const exportCSAFDocument = () => {
    const csafDocument = createCSAFDocument(documentStore)

    download(
      `${getFilename(documentStore.documentInformation.id)}.json`,
      JSON.stringify(csafDocument, null, 2),
    )
  }

  return { exportCSAFDocument }
}
