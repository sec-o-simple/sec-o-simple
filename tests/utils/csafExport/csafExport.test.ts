import { describe, it, expect, vi } from 'vitest'
import { createCSAFDocument } from '../../../src/utils/csafExport/csafExport'
import type { TDocumentStore } from '../../../src/utils/useDocumentStore'

// Mock the current date to ensure consistent snapshots
vi.setSystemTime(new Date('2023-07-31T12:00:00.000Z'))

describe('csafExport', () => {
  const createMockDocumentStore = (): TDocumentStore => ({
    sosDocumentType: 'Software',
    setSOSDocumentType: vi.fn(),
    documentInformation: {
      id: 'test-document-id',
      title: 'Test Security Advisory',
      lang: 'en',
      status: 'final',
      revisionHistory: [
        {
          date: '2023-01-01T00:00:00.000Z',
          number: '1.0.0',
          summary: 'Initial release',
        },
        {
          date: '2023-02-01T00:00:00.000Z',
          number: '1.1.0',
          summary: 'Security update',
        },
      ],
      tlp: {
        label: 'white',
        url: 'https://example.com/tlp',
      },
      publisher: {
        category: 'vendor',
        contactDetails: 'security@example.com',
        issuingAuthority: 'Security Team',
        name: 'Test Publisher',
        namespace: 'https://example.com',
      },
      references: [
        {
          summary: 'Vendor Advisory',
          url: 'https://example.com/advisory',
          category: 'external',
        },
      ],
      acknowledgments: [
        {
          organization: 'Security Research Org',
          names: [{ name: 'Alice Security' }, { name: 'Bob Researcher' }],
          summary: 'Discovered the vulnerability',
          url: 'https://example.com/researchers',
        },
      ],
      notes: [
        {
          id: 'note1',
          category: 'description',
          content: 'This is a test security advisory',
          title: 'Description',
        },
      ],
    },
    products: {
      'vendor-1': {
        id: 'vendor-1',
        category: 'vendor',
        name: 'Test Vendor',
        description: 'A test vendor for security testing',
        subBranches: [
          {
            id: 'product-1',
            category: 'product_name',
            name: 'Test Application',
            description: 'A sample test application for security testing',
            subBranches: [
              {
                id: 'version-1',
                category: 'product_version',
                name: '1.0.0',
                description: 'Version 1.0.0 of the test application',
                subBranches: [],
                type: 'Software',
              },
              {
                id: 'version-2',
                category: 'product_version',
                name: '2.0.0',
                description: 'Version 2.0.0 of the test application',
                subBranches: [],
                type: 'Software',
              },
            ],
            type: 'Software',
          },
        ],
      },
    },
    relationships: [
      {
        id: 'rel-1',
        category: 'default_component_of',
        productId1: 'product-1',
        productId2: 'linux-os',
        product1VersionIds: ['version-1'],
        product2VersionIds: ['linux-v1'],
        name: 'Test Application v1.0.0 on Linux',
      },
    ],
    vulnerabilities: {
      'vuln-1': {
        id: 'vuln-1',
        cve: 'CVE-2023-12345',
        title: 'Cross-site Scripting Vulnerability',
        cwe: {
          id: 'CWE-79',
          name: 'Improper Neutralization of Input During Web Page Generation',
        },
        notes: [
          {
            id: 'vuln-note-1',
            category: 'description',
            content: 'A cross-site scripting vulnerability exists in the application',
            title: 'Vulnerability Description',
          },
        ],
        products: [
          {
            id: 'vuln-product-1',
            status: 'known_affected',
            versions: [{ id: 'version-1', name: '1.0.0' }],
          },
          {
            id: 'vuln-product-2',
            status: 'fixed',
            versions: [{ id: 'version-2', name: '2.0.0' }],
          },
        ],
        remediations: [
          {
            category: 'mitigation',
            date: '2023-02-15T00:00:00.000Z',
            details: 'Update to version 2.0.0 or later',
            url: 'https://example.com/update-guide',
            productIds: ['version-1'],
          },
        ],
        scores: [
          {
            id: 'score-1',
            cvssVersion: '3.1',
            vectorString: 'CVSS:3.1/AV:N/AC:L/PR:N/UI:R/S:C/C:L/I:L/A:N',
            productIds: ['version-1'],
          },
          {
            id: 'score-2',
            cvssVersion: '4.0',
            vectorString: 'CVSS:4.0/AV:N/AC:L/AT:N/PR:N/UI:R/VC:L/VI:L/VA:N/SC:L/SI:L/SA:N',
            productIds: ['version-1'],
          },
        ],
      },
    },
    importedCSAFDocument: {},
    setImportedCSAFDocument: vi.fn(),
    updateDocumentInformation: vi.fn(),
    updateProducts: vi.fn(),
    updateRelationships: vi.fn(),
    updateVulnerabilities: vi.fn(),
    reset: vi.fn(),
  } as any)

  it('should create a CSAF document with complete data', () => {
    const mockStore = createMockDocumentStore()
    const mockConfig = { 
      template: {},
      productDatabase: { enabled: false }
    }
    
    const result = createCSAFDocument(mockStore, mockConfig)
    
    expect(result).toMatchSnapshot()
  })

  it('should create a CSAF document with minimal data', () => {
    const minimalStore = createMockDocumentStore()
    
    // Override with minimal data
    minimalStore.documentInformation = {
      ...minimalStore.documentInformation,
      revisionHistory: [],
      references: [],
      acknowledgments: [],
      notes: [],
      tlp: undefined,
    }
    minimalStore.products = {} as any
    minimalStore.relationships = []
    minimalStore.vulnerabilities = {} as any
    
    const result = createCSAFDocument(minimalStore, { 
      template: {},
      productDatabase: { enabled: false }
    })
    
    expect(result).toMatchSnapshot()
  })

  it('should handle vulnerabilities without optional fields', () => {
    const storeWithBasicVuln = createMockDocumentStore()
    
    storeWithBasicVuln.vulnerabilities = {
      'basic-vuln': {
        id: 'basic-vuln',
        cve: undefined,
        title: 'Basic Vulnerability',
        cwe: undefined,
        notes: [],
        products: [],
        remediations: undefined,
        scores: [],
      },
    } as any
    
    const result = createCSAFDocument(storeWithBasicVuln, { 
      template: {},
      productDatabase: { enabled: false }
    })
    
    expect(result).toMatchSnapshot()
  })
})
