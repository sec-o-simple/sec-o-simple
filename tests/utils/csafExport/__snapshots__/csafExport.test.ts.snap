// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`csafExport > should create a CSAF document with complete data 1`] = `
{
  "document": {
    "acknowledgments": [
      {
        "names": [
          "Alice Security",
          "Bob Researcher",
        ],
        "organization": "Security Research Org",
        "summary": "Discovered the vulnerability",
        "urls": [
          "https://example.com/researchers",
        ],
      },
    ],
    "category": "csaf_security_advisory",
    "csaf_version": "2.0",
    "distribution": {
      "tlp": {
        "label": "WHITE",
        "url": "https://www.first.org/tlp/",
      },
    },
    "lang": "en",
    "notes": [
      {
        "category": "description",
        "text": "This is a test security advisory",
        "title": "Description",
      },
      {
        "category": "description",
        "text": "A sample test application for security testing",
        "title": "Product description for Test Application",
      },
    ],
    "publisher": {
      "category": "vendor",
      "contact_details": "security@example.com",
      "issuing_authority": "Security Team",
      "name": "Test Publisher",
      "namespace": "https://example.com",
    },
    "references": [
      {
        "category": "external",
        "summary": "Vendor Advisory",
        "url": "https://example.com/advisory",
      },
    ],
    "title": "Test Security Advisory",
    "tracking": {
      "current_release_date": "2023-02-01T00:00:00.000Z",
      "generator": {
        "date": "2023-07-31T12:00:00.000Z",
        "engine": {
          "name": "Sec-O-Simple",
          "version": "0.0.1",
        },
      },
      "id": "test-document-id",
      "initial_release_date": "2023-01-01T00:00:00.000Z",
      "revision_history": [
        {
          "date": "2023-01-01T00:00:00.000Z",
          "number": "1.0.0",
          "summary": "Initial release",
        },
        {
          "date": "2023-02-01T00:00:00.000Z",
          "number": "1.1.0",
          "summary": "Security update",
        },
      ],
      "status": "final",
      "version": "1.1.0",
    },
  },
  "product_tree": {
    "branches": [
      {
        "branches": [
          {
            "branches": [
              {
                "category": "product_version",
                "name": "1.0.0",
                "product": {
                  "name": "Test Vendor Test Application 1.0.0",
                  "product_id": "version-1",
                },
              },
              {
                "category": "product_version",
                "name": "2.0.0",
                "product": {
                  "name": "Test Vendor Test Application 2.0.0",
                  "product_id": "version-2",
                },
              },
            ],
            "category": "product_name",
            "name": "Test Application",
          },
        ],
        "category": "vendor",
        "name": "Test Vendor",
      },
    ],
    "relationships": [
      {
        "category": "default_component_of",
        "full_product_name": {
          "name": "Test Vendor Test Application 1.0.0 default component of linux-v1",
          "product_id": "rel-1",
        },
        "product_reference": "version-1",
        "relates_to_product_reference": "linux-v1",
      },
    ],
  },
  "vulnerabilities": [
    {
      "cve": "CVE-2023-12345",
      "cwe": {
        "id": "CWE-79",
        "name": "Improper Neutralization of Input During Web Page Generation",
      },
      "flags": undefined,
      "notes": [
        {
          "category": "description",
          "text": "A cross-site scripting vulnerability exists in the application",
          "title": "Vulnerability Description",
        },
      ],
      "product_status": {
        "fixed": [
          undefined,
        ],
        "known_affected": [
          undefined,
        ],
      },
      "references": [
        {
          "category": "external",
          "summary": "CVSS v4.0 Score",
          "url": "https://www.first.org/cvss/calculator/4-0#CVSS:4.0/AV:N/AC:L/AT:N/PR:N/UI:R/VC:L/VI:L/VA:N/SC:L/SI:L/SA:N",
        },
      ],
      "remediations": [
        {
          "category": "mitigation",
          "date": "2023-02-15T00:00:00.000Z",
          "details": "Update to version 2.0.0 or later",
          "product_ids": [
            "version-1",
          ],
          "url": "https://example.com/update-guide",
        },
      ],
      "scores": [
        {
          "cvss_v3": {
            "baseScore": 6.1,
            "baseSeverity": "MEDIUM",
            "vectorString": "CVSS:3.1/AV:N/AC:L/PR:N/UI:R/S:C/C:L/I:L/A:N",
            "version": "3.1",
          },
          "products": [
            "version-1",
          ],
        },
      ],
      "title": "Cross-site Scripting Vulnerability",
    },
  ],
}
`;

exports[`csafExport > should create a CSAF document with minimal data 1`] = `
{
  "document": {
    "acknowledgments": undefined,
    "category": "csaf_security_advisory",
    "csaf_version": "2.0",
    "distribution": undefined,
    "lang": "en",
    "notes": undefined,
    "publisher": {
      "category": "vendor",
      "contact_details": "security@example.com",
      "issuing_authority": "Security Team",
      "name": "Test Publisher",
      "namespace": "https://example.com",
    },
    "references": undefined,
    "title": "Test Security Advisory",
    "tracking": {
      "current_release_date": "2023-07-31T12:00:00.000Z",
      "generator": {
        "date": "2023-07-31T12:00:00.000Z",
        "engine": {
          "name": "Sec-O-Simple",
          "version": "0.0.1",
        },
      },
      "id": "test-document-id",
      "initial_release_date": "2023-07-31T12:00:00.000Z",
      "revision_history": [],
      "status": "final",
      "version": "1",
    },
  },
  "product_tree": {
    "branches": [],
    "relationships": undefined,
  },
  "vulnerabilities": [],
}
`;

exports[`csafExport > should handle vulnerabilities without optional fields 1`] = `
{
  "document": {
    "acknowledgments": [
      {
        "names": [
          "Alice Security",
          "Bob Researcher",
        ],
        "organization": "Security Research Org",
        "summary": "Discovered the vulnerability",
        "urls": [
          "https://example.com/researchers",
        ],
      },
    ],
    "category": "csaf_security_advisory",
    "csaf_version": "2.0",
    "distribution": {
      "tlp": {
        "label": "WHITE",
        "url": "https://www.first.org/tlp/",
      },
    },
    "lang": "en",
    "notes": [
      {
        "category": "description",
        "text": "This is a test security advisory",
        "title": "Description",
      },
      {
        "category": "description",
        "text": "A sample test application for security testing",
        "title": "Product description for Test Application",
      },
    ],
    "publisher": {
      "category": "vendor",
      "contact_details": "security@example.com",
      "issuing_authority": "Security Team",
      "name": "Test Publisher",
      "namespace": "https://example.com",
    },
    "references": [
      {
        "category": "external",
        "summary": "Vendor Advisory",
        "url": "https://example.com/advisory",
      },
    ],
    "title": "Test Security Advisory",
    "tracking": {
      "current_release_date": "2023-02-01T00:00:00.000Z",
      "generator": {
        "date": "2023-07-31T12:00:00.000Z",
        "engine": {
          "name": "Sec-O-Simple",
          "version": "0.0.1",
        },
      },
      "id": "test-document-id",
      "initial_release_date": "2023-01-01T00:00:00.000Z",
      "revision_history": [
        {
          "date": "2023-01-01T00:00:00.000Z",
          "number": "1.0.0",
          "summary": "Initial release",
        },
        {
          "date": "2023-02-01T00:00:00.000Z",
          "number": "1.1.0",
          "summary": "Security update",
        },
      ],
      "status": "final",
      "version": "1.1.0",
    },
  },
  "product_tree": {
    "branches": [
      {
        "branches": [
          {
            "branches": [
              {
                "category": "product_version",
                "name": "1.0.0",
                "product": {
                  "name": "Test Vendor Test Application 1.0.0",
                  "product_id": "version-1",
                },
              },
              {
                "category": "product_version",
                "name": "2.0.0",
                "product": {
                  "name": "Test Vendor Test Application 2.0.0",
                  "product_id": "version-2",
                },
              },
            ],
            "category": "product_name",
            "name": "Test Application",
          },
        ],
        "category": "vendor",
        "name": "Test Vendor",
      },
    ],
    "relationships": [
      {
        "category": "default_component_of",
        "full_product_name": {
          "name": "Test Vendor Test Application 1.0.0 default component of linux-v1",
          "product_id": "rel-1",
        },
        "product_reference": "version-1",
        "relates_to_product_reference": "linux-v1",
      },
    ],
  },
  "vulnerabilities": [
    {
      "cve": undefined,
      "cwe": undefined,
      "flags": undefined,
      "notes": [],
      "product_status": {},
      "references": undefined,
      "remediations": undefined,
      "scores": undefined,
      "title": "Basic Vulnerability",
    },
  ],
}
`;
